include/fi_mem.h

```diff
@@ -69,24 +69,31 @@ static inline void *mem_dup(const void *src, size_t size)
  */
 #define FREESTACK_EMPTY	NULL
 
+#define FREESTACK_HEADER 					\
+	size_t		size;					\
+	void		*next;					\
+
 #define freestack_isempty(fs)	((fs)->next == FREESTACK_EMPTY)
 #define freestack_push(fs, p)					\
-{								\
+do {								\
 	*(void **) p = (fs)->next;				\
 	(fs)->next = p;						\
+} while (0)
+#define freestack_pop(fs) freestack_pop_impl(fs, fs->next)
```

Comment:
I think `fs` should be surrounded in parenthesis before accessing the `next` field.